traits/observers/expression.py

```diff
@@ -104,6 +110,33 @@ def then(self, expression):
             new._prior_expression = _SeriesExpression([self, expression])
         return new
 
+    def trait(self, name, notify=True, optional=False):
+        """ Create a new expression for observing a trait with the exact
+        name given.
+
+        e.g. ``trait("child").trait("age")`` matches ``child.age``
+        on an object, and is equivalent to
+        ``trait("child").then(trait("age"))``
```

Comment:
Could we stick with just one way to do it here? I think `trait("child").then(trait("age"))` should be fine for now; we could add the `trait("child").trait("age")` support later if we think it's needed.

In some sense, this is a violation of the OCP: `then` is a general piece of machinery that lets us chain expressions, while if we follow this pattern we'll want an extra method for each new type of observer - we end up repeating ourselves, and each new observer type expects a new method in this class.

Reply:
Ah, I see, there isn't a repetition here, because of the `_as_top_level` magic that converts the method into a matching function. I'd recommend removing the magic and the method, and having a direct top-level function for creating the expression. That'll make for simpler, more direct code that's easier to maintain going forward.

Reply:
This is very much a user interface (just not graphical or command-line-like), a bit like a main application.  It seems laborious to do `trait("child").then(trait("attr1")).then(trait("attr2")).then(trait("attr3"))` when we could do `trait("child").trait("attr").trait("attr2"))`

Likewise, this is more user-friendly:
`trait("child").metadata("name")`
than
`trait("child").then(metadata("name"))`
The latter also requires importing `metadata` in addition to `trait`.

Reply:
@corranwebster Since you have reviewed the user experience before, thought I should check again if we are changing this. Is it okay if we ditch `trait("name1").trait("name2").trait("name3")` but require the user to go with `trait("name1").then(trait("name2")).then(trait("name3"))` instead?

Reply:
(This will apply to all the other features, e.g. `metadata` and `filter_`)

Reply:
Okay. Then it may still be worth changing the logic so that in this `trait` _method_ we simply do `return self.then(trait(name, ...))`, and the `trait` function is responsible for returning a simple `_NamedTraitObserver` expression.

It still feels backwards that the way to create a simple Expression containing a `_NamedTraitObserver` is to create an empty `Expression` and then chain something to it. I'd like to fix that before the release, and I think it'll be easier to fix if the logic is encapsulated in the top-level function. I'm also not fond of the decorator magic - it makes the code again harder to understand, maintain, and debug, and I don't think we need it.

Reply:
I did consider the other way round: Have the top-level function doing the actual work and then the method calls the top-level function instead.

In that case, we will have to maintain two sets of docstrings and signatures for every single element of the expression (justified by this being more user-friendly). So there is additional maintenance cost in that direction too. In other words, it is a tradeoff, and the weight for which is stronger may not be quantifiable easily.

Anyway, I am happy to go this other route too.

Reply:
I am fine with this sort of change to the API.

Reply:
OK!

Reply:
(This means a lot of the tests I have written for future PRs will need to change. Sigh.)

Reply:
I'm fine with keeping the `.trait` method, FWIW, just so long as we untangle so that it's clear that it's the `.trait` method that depends on the function, and not the other way around.

Reply:
Thanks. I will keep that then.